File: src/main/scala/temportalist/compression/main/common/block/tile/TileCompressedTickable.scala
package temportalist.compression.main.common.block.tile

import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.{Entity, EntityLivingBase, SharedMonsterAttributes}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.{AxisAlignedBB, BlockPos}
import net.minecraft.util.{DamageSource, ITickable}
import temportalist.compression.main.common.init.ModBlocks
import temportalist.compression.main.common.item.ItemDenseArmor
import temportalist.compression.main.common.lib.{EnumTier, Ticks}
import temportalist.compression.main.common.{Effects, Options}
import temportalist.origin.api.common.lib.Vect

import scala.collection.JavaConversions

/**
  *
  * Created by TheTemportalist on 4/16/2016.
  *
  * @author TheTemportalist
  */
class TileCompressedTickable extends TileCompressed with ITickable {

	private var posCenter: Vect = _

	private val blockDestroyDelay_base = Ticks.getTicks(seconds = 20)
	private val blockDestroyDelay_deviation = Ticks.getTicks(seconds = 2)
	private var blockDestroyDelay_till = -1
	private var blockDestroyDelay_needsReset = true
	private val blockDestroy_radius = 5

	private val entityRadius = 10
	private var entityAABB: AxisAlignedBB = _

	private val checkEnergyBounds_base = Ticks.getTicks(minutes = 4)
	private var checkEnergyBounds_till = -1
	private var checkEnergyBounds_needsReset = true
	private var storedEnergy: Double = 0
	private val energyRadius = 1.5D
	private var energyAABB: AxisAlignedBB = _

	override def update(): Unit = {
		if (EnumTier.getTierForSize(this.getSize).ordinal() + 1 < Options.blackHole) return

		if (this.posCenter == null) this.posCenter = new Vect(this) + Vect.CENTER

		this.updateBlockDestroyLogic()
		this.attractEntities()

		if (Options.blackHolePotentialEnergy)
			this.checkEnergyBoundsLogic()

	}

	private def updateBlockDestroyLogic(): Unit = {
		if (!this.blockDestroyDelay_needsReset && this.blockDestroyDelay_till <= 0) {
			this.destroyBlock()
			this.blockDestroyDelay_needsReset = true
		} else this.blockDestroyDelay_till -= 1
		this.blockDestroy_resetDelay()
	}

	private def blockDestroy_resetDelay(): Unit = {
		if (this.blockDestroyDelay_needsReset) {
			this.blockDestroyDelay_till = this.blockDestroyDelay_base +
					this.getWorld.rand.nextInt(this.blockDestroyDelay_deviation)
			this.blockDestroyDelay_till = 20
			this.blockDestroyDelay_needsReset = false
		}
	}

	private def destroyBlock(): Unit = {
		if (this.getWorld.isRemote) return

		val thisVect = new Vect(this)

		var vect: Vect = null
		var pos: BlockPos = null
		var state: IBlockState = null
		var loops = 0
		val loopMax = 100
		do {
			loops += 1
			vect = new Vect(this.getRandPosRadius, this.getRandPosRadius, this.getRandPosRadius)
			vect += thisVect
			pos = vect.toBlockPos
			state = this.getWorld.getBlockState(pos)
		}
		while (loops < loopMax && !this.isValidStateToDestroy(vect, pos, state))

		if (loops > loopMax) return

		this.getWorld.setBlockToAir(pos)

	}

	private def getRandPosRadius: Int =
		this.getWorld.rand.nextInt(blockDestroy_radius * 2 + 1) - this.blockDestroy_radius

	private def isValidStateToDestroy(vect: Vect, pos: BlockPos, state: IBlockState): Boolean = {
		pos != this.getPos && vect.length <= this.blockDestroy_radius &&
				!this.isInDestroyBlacklist(state) &&
				state.getMaterial != Material.AIR && !state.getMaterial.isLiquid
	}

	private def isInDestroyBlacklist(state: IBlockState): Boolean = {
		state.getBlock == ModBlocks.block && state.getBlock.hasTileEntity(state)
	}

	private def attractEntities(): Unit = {

		if (this.entityAABB == null) {
			val pos = this.getPos
			this.entityAABB = new AxisAlignedBB(
				pos.getX - this.entityRadius, pos.getY - this.entityRadius, pos.getZ - this.entityRadius,
				pos.getX + this.entityRadius, pos.getY + this.entityRadius, pos.getZ + this.entityRadius
			)
		}

		val entityList = this.getWorld.getEntitiesWithinAABB(classOf[Entity], this.entityAABB)
		for (entity <- JavaConversions.asScalaBuffer(entityList)) {
			if (this.shouldPullEntity(entity)) {
				var speed = 1D
				entity match {
					case player: EntityPlayer =>
						speed -= ItemDenseArmor.getClothedCount(player) * 0.25
					case _ =>
				}
				Effects.pullEntityTowards(entity, this.posCenter, Vect.ZERO, speed = speed)
			}
		}

	}

	private def shouldPullEntity(entity: Entity): Boolean = {
		entity match {
			case player: EntityPlayer => !player.capabilities.isCreativeMode
			case _ => true
		}
	}

	private def checkEnergyBoundsLogic(): Unit = {
		if (!this.checkEnergyBounds_needsReset && this.checkEnergyBounds_till <= 0) {
			this.checkEnergyBounds()
			this.checkEnergyBounds_needsReset = true
		} else this.checkEnergyBounds_till -= 1
		this.checkEnergy_resetDelay()
	}

	private def checkEnergy_resetDelay(): Unit = {
		if (this.checkEnergyBounds_needsReset) {
			this.checkEnergyBounds_till = this.checkEnergyBounds_base
			this.checkEnergyBounds_till = 20
			this.checkEnergyBounds_needsReset = false
		}
	}

	private def checkEnergyBounds(): Unit = {
		if (this.getWorld.isRemote) return

		if (this.energyAABB == null) {
			val pos = this.posCenter
			this.energyAABB = new AxisAlignedBB(
				pos.x - this.energyRadius, pos.y - this.energyRadius, pos.z - this.energyRadius,
				pos.x + this.energyRadius, pos.y + this.energyRadius, pos.z + this.energyRadius
			)
		}

		val entityList = this.getWorld.getEntitiesWithinAABB(classOf[Entity], this.energyAABB)
		for (entity <- JavaConversions.asScalaBuffer(entityList)) {
			entity match {
				case living: EntityLivingBase =>
					if (living.attackEntityFrom(DamageSource.outOfWorld, 1)) {
						if (living.getHealth <= 0) {
							living match {
								case player: EntityPlayer => this.giveEnergyTo(player)
								case _ => this.storedEnergy += living.getMaxHealth / 10
							}
						}
					}
				case entityItem: EntityItem =>
					this.storedEnergy += entityItem.getEntityItem.stackSize
					entityItem.setDead()
				case _ => entity.setDead()
			}
		}

	}

	private def giveEnergyTo(player: EntityPlayer): Unit = {
		val heartsToGrant = this.storedEnergy.toInt % 10
		player.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(
			player.getMaxHealth + 2 * heartsToGrant
		)
		this.storedEnergy -= heartsToGrant * 10
	}

	override def writeToNBT(compound: NBTTagCompound): NBTTagCompound = {
		val tag = super.writeToNBT(compound)
		tag.setInteger("blockDestroyDelay_till", this.blockDestroyDelay_till)
		tag.setInteger("checkEnergyBounds_till", this.checkEnergyBounds_till)
		tag.setDouble("storedEnergy", this.storedEnergy)
		tag
	}

	override def readFromNBT(compound: NBTTagCompound): Unit = {
		super.readFromNBT(compound)
		this.blockDestroyDelay_till = compound.getInteger("blockDestroyDelay_till")
		this.checkEnergyBounds_till = compound.getInteger("checkEnergyBounds_till")
		this.storedEnergy = compound.getDouble("storedEnergy")

	}

}

File: src/main/java/com/temportalist/compression/common/blocks/TileCompressedTickable.java
package com.temportalist.compression.common.blocks;

import com.temportalist.compression.common.effects.Effects;
import com.temportalist.compression.common.effects.EnumEffect;
import com.temportalist.compression.common.init.ModBlocks;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.DamageSource;
import net.minecraft.util.ITickable;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;

import java.util.List;
import java.util.Random;

public class TileCompressedTickable extends TileCompressed implements ITickable {

    static abstract class Ticker {

        private int ticksUntil;
        private final int delayRangeMin, delayRangeMax; // max is exclusive

        public Ticker(int minDelay, int maxDelay) {
            this.delayRangeMin = minDelay;
            this.delayRangeMax = maxDelay;
            this.ticksUntil = maxDelay;
        }

        public Ticker(int maxDelayInTicks) {
            this(0, maxDelayInTicks);
        }

        public void update(Random random) {

            // Check to see if effect can occur
            if (this.ticksUntil <= 0) {
                // Do the effect
                this.doEffect();
                // Reset the tick counter
                this.ticksUntil = this.getNextTicksUntil(random);
            } else {
                // Decrement the ticks
                this.ticksUntil -= 1;
            }

        }

        private int getNextTicksUntil(Random random) {
            return this.delayRangeMin == 0 ? this.delayRangeMax : random.nextInt(this.delayRangeMax - this.delayRangeMin) + this.delayRangeMin;
        }

        abstract public void doEffect();

    }

    // The absolute center of this blocktile
    private Vec3d center;

    // The ticker structures to track timing
    private Ticker tickerDestroyBlocks, tickerDamageEntities;
    // The bounding boxes for attracting entities
    private AxisAlignedBB aoeAttract, aoeDamage;
    // The amount of energy that has been accumulated
    private float energyAmount;

    // The radius from this blocks that the black hole can consume
    private final int radiusBlockDestroy = 5;
    // The radius from this block that the block can pull entities
    private final float radiusAttractEntities = 10;
    // The radius from this block that the block can damage entities
    private final float radiusDamage = 1.5f;

    public TileCompressedTickable() {
        this.center = new Vec3d(this.getPos()).add(new Vec3d(0.5, 0.5, 0.5));
        this.tickerDestroyBlocks = this.createTickerDestroyBlocks();
        this.tickerDamageEntities = this.createTickerDamageEntities();
        this.aoeAttract = new AxisAlignedBB(this.getPos()).grow(this.radiusAttractEntities);
        this.aoeDamage = new AxisAlignedBB(this.getPos()).grow(this.radiusDamage);
        this.energyAmount = 0;
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound compound) {
        NBTTagCompound tagCom = super.writeToNBT(compound);
        tagCom.setFloat("energyAmount", this.energyAmount);
        return tagCom;
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);
        this.energyAmount = compound.getFloat("energyAmount");
    }

    public float getEnergyAmount() {
        return energyAmount;
    }

    @Override
    public void update() {
        this.updateBlackHole();
    }

    private void updateBlackHole() {
        if (!EnumEffect.BLACK_HOLE.canDoEffect(this.getTier())) return;

        // Update ticking effects
        this.tickerDestroyBlocks.update(this.getWorld().rand);
        this.tickerDamageEntities.update(this.getWorld().rand);

        // Attract local entities
        this.attractEntities();

    }

    private Ticker createTickerDestroyBlocks() {

        final int maxLoops = 100;

        int delayBase = 20 * 20; // 20 seconds
        int deviation = 2 * 20; // 2 seconds
        return new Ticker(delayBase, delayBase + deviation) {

            @Override
            public void doEffect() {
                int loops = 0;
                BlockPos pos;
                IBlockState state;
                do {
                    pos = this.getRandBlockPos().add(getPos());
                    state = getWorld().getBlockState(pos);
                    loops++;
                } while (loops < maxLoops && !canBeDestroyed(state, pos));

                if (canBeDestroyed(state, pos)) {
                    consume(state, pos);
                    getWorld().setBlockToAir(pos);
                }

            }

            int getRandBlockCoord() {
                return getWorld().rand.nextInt(radiusBlockDestroy);
            }

            BlockPos getRandBlockPos() {
                //return BlockPos.ORIGIN.add(0, 1, 0);
                ////*
                return new BlockPos(
                        this.getRandBlockCoord(),
                        this.getRandBlockCoord(),
                        this.getRandBlockCoord()
                );
                //*/
            }

        };
    }

    private Ticker createTickerDamageEntities() {

        return new Ticker(30 * 20) { // do damage every 30 seconds
            @Override
            public void doEffect() {

                List<Entity> entities = getWorld().getEntitiesWithinAABB(Entity.class, aoeDamage,
                        (Entity entity) -> entity instanceof EntityItem || entity instanceof EntityLivingBase
                );
                for (Entity entity : entities) {
                    if (entity instanceof EntityItem) {
                        consume(((EntityItem) entity).getItem());
                    } else if (entity instanceof EntityLivingBase) {
                        EntityLivingBase living = (EntityLivingBase) entity;
                        if (living.attackEntityFrom(DamageSource.OUT_OF_WORLD, 1)) {
                            if (living.getHealth() <= 0) {
                                consume(living);
                            }
                        }
                    }
                }

            }
        };
    }

    public boolean canBeDestroyed(IBlockState state, BlockPos pos) {
        boolean thisPos = pos == this.getPos();
        boolean validDist = pos.distanceSq(this.getPos()) <= this.radiusBlockDestroy * this.radiusBlockDestroy;
        boolean validMaterial = state.getMaterial() != Material.AIR && !state.getMaterial().isLiquid();
        boolean validTile = !state.getBlock().hasTileEntity(state) && getWorld().getTileEntity(pos) == null;

        return !thisPos && validDist && validMaterial && validTile;
    }

    public void consume(IBlockState state, BlockPos pos) {
        if (!EnumEffect.POTENTIAL_ENERGY.canDoEffect(this.getTier())) return;
        this.gainEnergy(1);
    }

    public void consume(ItemStack stack) {
        if (!EnumEffect.POTENTIAL_ENERGY.canDoEffect(this.getTier())) return;
        this.gainEnergy(stack.getCount());
    }

    public void consume(EntityLivingBase entity) {
        if (!EnumEffect.POTENTIAL_ENERGY.canDoEffect(this.getTier())) return;
        this.gainEnergy(entity.getMaxHealth());
    }

    public void gainEnergy(float amount) {
        this.energyAmount += amount;
        ModBlocks.compressed.updateBlockStats(
                this.getWorld(), this.getPos(), this.getTier(), this.getSampleStack(), this.energyAmount);
    }

    public void attractEntities() {

        List<Entity> entities = this.getWorld().getEntitiesWithinAABB(Entity.class, this.aoeAttract,
                (Entity entity) -> {
                    if (entity instanceof EntityPlayer && ((EntityPlayer) entity).isCreative()) return false;
                    else return entity != null && !entity.getEntityBoundingBox().intersects(this.aoeDamage);
                }
        );
        for (Entity entity : entities) {
            Effects.pullEntityTowards(entity, this.center, 1);
        }

    }

}
